057-marking-zk-kraft-clusters.md

```diff
@@ -0,0 +1,59 @@
+# Marking ZooKeeper and KRaft based clusters for KRaft enabled operators
+
+This proposal is about changing the way the current `UseKRaft` feature gate works.
+It changes the way the Strimzi Cluster Operator handles ZooKeeper and KRaft based clusters when such feature gate is enabled.
+
+## Current situation
+
+Currently, when the `UseKRaft` feature gate is enabled (together with the required `KafkaNodePools` one), the operator expects all the already running Apache Kafka clusters being KRaft-based.
+It means that the `Kafka` custom resources are configured together with `KafkaNodePool`(s) with `broker` and `controller` roles.
+There is no way to differentiate between clusters running in ZooKeeper or KRaft mode.
+When a `Kafka` custom resource is configured to actually use ZooKeeper or it's just badly configured but supposed to be KRaft based, the operator detects it as having a missing KRaft controllers configuration and logs the following warning:
+
+```shell
+io.strimzi.operator.common.model.InvalidResourceException: Tha Kafka cluster my-cluster is invalid: [At least one KafkaNodePool with the controller role and at least one replica is required when KRaft mode is enabled]
+```
+
+In this case, the reconciliation fails and the cluster is not operated at all.
+It means that the operator doesn't take any actions on changes applied by the user to the `spec.kafka` or `spec.zookeeper` in the `Kafka` custom resource.
+
+## Motivation
+
+Leaving a ZooKeeper-based cluster not operated when the `UseKRaft` feature gate is enabled looks like a bad thing for the user, until the ZooKeeper support will be completely removed.
+Before that, the user should be able to have both ZooKeeper and KRaft based clusters running and operated when the `UseKRaft` feature gate is enabled.
+Also, taking into account the future ZooKeeper to KRaft migration process, as described in the proposal [PR#90](https://github.com/strimzi/proposals/pull/90), it would help the operator migration component to detect which cluster is ZooKeeper-based so "migrate-able" and which one is already KRaft-based.
+
+## Proposal
+
+The proposal is about adding a new `strimzi.io/kraft` annotation to the `Kafka` custom resource, to be applied by the user, in order to identify a ZooKeeper or KRaft based cluster.
+The possible values would be:
+
+* `disabled` (or missing): identifies a ZooKeeper-based cluster.
```

Comment:
I guess any invalid value should default to ZooKeeper as well. That way you can easily handle it with `if ("enabled".equals(annoValue)) { ... }`.